### DISCUSSION TITLE
How can I use transaction using this library

### DISCUSSION THREAD

**Discussion by jollyamit** [Havret/dotnet-activemq-artemis-client#257] [Q&A] [answered]
How can I use transaction using this library?
My use case is as below.
1) Create message
2) Send message to Artemis queue 
3) do something else. (This step result processing failed or success) 
4) Based upon outcome of step 3, I need to either commit or rollback the message sent in step 2.

**Comment by Havret** [comment, marked as answer]
@jollyamit The simplest example may look as follows:

```csharp
using System;
using System.Threading.Tasks;
using ActiveMQ.Artemis.Client;
using ActiveMQ.Artemis.Client.Transactions;

namespace Example
{
    class Program
    {
        static async Task Main()
        {
            var endpoint = Endpoint.Create("localhost", 5672, "guest", "guest");
            var connectionFactory = new ConnectionFactory();
            await using var connection = await connectionFactory.CreateAsync(endpoint);
            await using var producer = await connection.CreateProducerAsync("my-address");
            
            await using var transaction = new Transaction();
            var message = new Message("my-payload");
            await producer.SendAsync(message, transaction);

            var result = DoSthThatMayFail();
            if (result)
                await transaction.CommitAsync();
            else
                await transaction.RollbackAsync();

        }

        private static bool DoSthThatMayFail()
        {
            var random = new Random();
            return random.Next() % 2 == 0;
        }
    }
}
```

You can find more examples in [`TransactionsSpec.cs`](https://github.com/Havret/dotnet-activemq-artemis-client/blob/85c7ea3c90d8a0758481583b0c87fa9dc93cc398/test/ActiveMQ.Artemis.Client.IntegrationTests/TransactionsSpec.cs).

**Reply by jollyamit** [reply]
Thx Havret. I guess I should be able to use the same way on the consumer side too.?

**Reply by Havret** [reply]
@jollyamit Yes, it will be similar on the consuming side. You can as well combine the two together.

The typical flow is that you receive a message [1], do some transformation and publish another message [2] as a result. In order to be sure that the [1] will be acknowledge only if you successfully sent the [2], you can do sth like this:

```csharp
using System;
using System.Threading.Tasks;
using ActiveMQ.Artemis.Client;
using ActiveMQ.Artemis.Client.Transactions;

namespace Example
{
    class Program
    {
        static async Task Main()
        {
            var endpoint = Endpoint.Create("localhost", 5672, "guest", "guest");
            var connectionFactory = new ConnectionFactory();
            await using var connection = await connectionFactory.CreateAsync(endpoint);
            await using var producer = await connection.CreateProducerAsync("my-address-2", RoutingType.Anycast);
            await using var consumer = await connection.CreateConsumerAsync("my-address-1", RoutingType.Anycast);

            var msg1 = await consumer.ReceiveAsync();

            await using (var transaction = new Transaction())
            {
                // transactionally acknowledge
                await consumer.AcceptAsync(msg1, transaction);

                var result = DoSomeProcessing(msg1);
                var msg2 = new Message(result);

                // transactionally send another message
                await producer.SendAsync(msg2, transaction);

                // commit two operations
                await transaction.CommitAsync();
            }
        }

        private static string DoSomeProcessing(Message message)
        {
            var payload = message.GetBody<string>();
            var random = new Random();
            return $"{payload}-{random.Next()}";
        }
    }
}
```

**Reply by jollyamit** [reply]
Thanks @Havret for quick response it really helps. 

On the different topic, how can I setup a DurableSubscriptions as explained in the example below.

https://github.com/apache/activemq-artemis/blob/master/examples/protocols/amqp/dotnet/DurableSubscriptions/Program.cs

I guess basically I need to provide the Container Id while opening the connection and provide the Subscription name while opening the ReceiverLink.

Connection connection = new Connection(new Address(DEFAULT_BROKER_URI), 
                                                   SaslProfile.Anonymous,
                                                   new Open() { ContainerId = "My-Container-1" }, null);

ReceiverLink receiver = new ReceiverLink(session, "My-Subscription-1", source, null);

I think in ConnectionBuilder.cs we need a way to pass the Ampq.Framing.Open object and similarly in ConsumerBuilder.cs we need a way to pass SubscriptionId (If null we can set to GUID)

https://github.com/Havret/dotnet-activemq-artemis-client/blob/master/src/ActiveMQ.Artemis.Client/Builders/ConnectionBuilder.cs

var connection = await connectionFactory.CreateAsync(endpoint.Address, null, OnOpened).ConfigureAwait(false);

https://github.com/Havret/dotnet-activemq-artemis-client/blob/master/src/ActiveMQ.Artemis.Client/Builders/ConsumerBuilder.cs

var receiverLink = new ReceiverLink(_session, Guid.NewGuid().ToString(), source, OnAttached);

**Reply by Havret** [reply]
@jollyamit Durable subscriptions as in your example are not supported by design, because they may cause numerous problems. Moreover, they are hardly ever needed. The only scenario I can think of is when you try to build your own flavor of `distributed` cache. You have multiple consumers and each one needs to get its own copy of the message. Even so, it can be done without durable subscriptions. You can find an example of reliable pub-sub implementation here. --> https://github.com/Havret/activemq-artemis-net-reliable-pub-sub

But I bet (based on the fact that you're asking about transactions) that this is not what you are looking for. :) If you could describe your use case I would probably provide better suggestions.

**Reply by jollyamit** [reply]
@Havret my use case is, I want to achieve JMS Pub/Sub kind of setup with an option of both Durable and Non Durable subscriptions i.e. client, when connecting to a Topic (a.k.a Multicast Address) should be able to dynamically specify durability of its subscription. Durable means even if particular consumer is down, message should be accumulated on server and should be delivered when it comes back online with same client-id/subscription-id. I think I should be able to achieve the expected behavior as provided in your example by setting AutoDelete = false in the QueueConfiguration?

**Reply by Havret** [reply]
@jollyamit Yes, that's correct. It will work as you described. Just be mindful that with this particular setup, when your consumer goes down, but the producer keeps pushing new messages, it may kill the broker. A more reliable solution is to create *a shared subscription*. In this case, you create a well-known address and well-known queue. For instance:

```
- my address (Multicast)
  - my-queue (Multicast)
     - consumer1 
     - consumer2
```

Now you have two (or more) competing consumers that can pick up messages from the queue in a round-robin fashion, and when one goes down, the other will take care of all the messages. When it's back the load will be distributed evenly again. This is the topology I am using in 99% of my use cases.

**Reply by jollyamit** [reply]
@Havret, My use case is, both consumer1 and consumer2 needs to get all the messages published on a given address/topic.

Let's say if publisher publishes 1000000 messages in a day, I want all 1000000 messages to be consumed by both consumer1 and consumer2 regardless if they are online or offline. Idea is they both should get 1000000 messages if they are online or when ever they come online. (Under normal circumstance, they both should be online, but very rare case there might be possibility that one of the them or both are offline for a while) 

Also I need an ability to have durable consumer subscription created dynamically/programmatically.

Based upon your comment above, if producer is up and keep pushing the messages on Topic, and let's say one of the durable consumers with subscription id (subscription-1) goes down for lets say 1 hour, and other durable consumers is still UP, why would it create the problem for broker? All the messages will be written do the disk on broker side, broker will not hold everything for durable consumers 1 with subscription-1 in the memory which is currently down... correct?  Once consumers 1 with subscription-1  come back online again, broker will deliver all the messages to it.

In your example below, I tested and noticed below 
1) Messages remains in the my-queue even if both consumer1 (connecting to my-address) & consumer1 (connecting to my-address) consumes the messages and commit the transaction. ( I think this is expected behavior? as consumer are not directly connecting to my-queue either using   my-address::my-queue or using my-queue as subscription id ?)
2) Broker will not deliver the messages it received during the time consumer were down but it came back up lets say after 20 mins. (Messages received by broker during those 20 mins are not delivered back to consumer when it came back online.)
 
- my address (Multicast)
  - my-queue (Multicast)
     - consumer1 
     - consumer2

**Reply by Havret** [reply]
>Let's say if publisher publishes 1000000 messages in a day, I want all 1000000 messages to be consumed by both consumer1 and consumer2 regardless if they are online or offline. Idea is they both should get 1000000 messages if they are online or when ever they come online. (Under normal circumstance, they both should be online, but very rare case there might be possibility that one of the them or both are offline for a while)

In this case I would create the following topology:
```
- my address (Multicast)
  - my-queue1 (Multicast)
     - consumer(1) (service1)
     ...
     - consumer(n) (service1)
  - my-queue2 (Multicast)
     - consumer(1) (service2)
     ...
     - consumer(n) (service2)
```    

A single multicast address, and two queues under this address for two types of your consumers (service1, service2). In this configuration you have two different queues under the same address, so the broker will deliver all messages to my-queue1 and my-queue2. You can attach as many consumers as you like to these queues (scale out) to ensure high availability. If you have a requirement that only a single consumer of each type should be able to process messages, you can mark the queue as Exclusive. The broker will pick a single consumer to get all the messages from a queue. If that consumer fails, the broker will auto failover and choose another consumer. For most of my use cases I just run many parallel consumers [with message grouping enabled](https://activemq.apache.org/components/artemis/documentation/latest/message-grouping.html).

An example implementation may look like this --> https://github.com/Havret/activemq-artemis-net-example

> Based upon your comment above, if producer is up and keep pushing the messages on Topic, and let's say one of the durable consumers with subscription id (subscription-1) goes down for lets say 1 hour, and other durable consumers is still UP, why would it create the problem for broker? All the messages will be written do the disk on broker side, broker will not hold everything for durable consumers 1 with subscription-1 in the memory which is currently down... correct? Once consumers 1 with subscription-1 come back online again, broker will deliver all the messages to it.

This is not that simple. Please check this out --> https://activemq.apache.org/components/artemis/documentation/latest/paging.html

**Reply by jollyamit** [reply]
Thanks @Havret for detailed explanation. Basically in your above example my-queue1 & my-queue2 are nothing but a dynamically created queues by the below mentioned DurableSubscriptions example program. In guess with your Topology manager class same behavior can be achieve. As you explained in https://github.com/Havret/activemq-artemis-net-reliable-pub-sub example with AutoDelete = false in the QueueConfiguration.

https://github.com/apache/activemq-artemis/blob/master/examples/protocols/amqp/dotnet/DurableSubscriptions/Program.cs

**Reply by Havret** [reply]
Basically yes, the difference is that there is no static `connectionId/containerId` which may cause problems during connection recovery, and of course you have full control over how the queue is configured.

**Reply by jollyamit** [reply]
Correct.. That's why I was thinking to add a support in your lib to be able to provide connectionId/containerId instead of GUID :)
